pkg/trtl/trtl.go

```diff
@@ -1,156 +1,566 @@
+	// NOTE: this section must come after the iterator is created, though it would be preferable
+	// if it was in the section where the page cursor is created. This is because we want to
+	// check the namespace that the iterator is operating on rather than the one specified by
+	// the user in the first request (e.g. because a default namespace may be used).
+	if opts.PageToken != "" && cursor.Namespace != iter.Namespace() {
+		log.Debug().Msg("invalid iter request: mismatched namespace")
+		return nil, status.Error(codes.InvalidArgument, "namespace cannot change between requests")
+	}
+
+	// If necessary seek to the next key specified by the cursor.
+	if len(cursor.NextKey) > 0 {
+		// If iter.Seek returns false (e.g. seek did not find the specified key) then
+		// iter.Next() should also return false, so it isn't necessary to check the return.
+		// NOTE: next key must be set to nil after it's used for seeking so that the last
+		// page doesn't retain the old key and loop forever.
+		iter.Seek(cursor.NextKey)
+		cursor.NextKey = nil
+
+		// Because we're going to be calling Next, we need to back up one key to ensure
+		// that we start on the right key in the for loop.
+		iter.Prev()
+	}
+
+	for iter.Next() {
+		// Check if we're done iterating (e.g. at the end of the page with a next page)
+		if len(out.Values) == int(cursor.PageSize) {
+			// The current key is the next key for the next page, stop iteration and
+			// prepare the page cursor to be returned.
+			cursor.NextKey = iter.Key()
+			cursor.Namespace = iter.Namespace()
+			break
+		}
+
+		// Otherwise append the current key value pair to the page.
+		// Fetch the metadata since it will need to be loaded for the response anyway.
+		var object *object.Object
+		if object, err = iter.Object(); err != nil {
+			log.Error().Err(err).Str("key", base64.RawURLEncoding.EncodeToString(iter.Key())).Msg("could not fetch object metadata")
+			return nil, status.Error(codes.FailedPrecondition, "database is in invalid state")
+		}
+
+		// Create the key value pair
+		pair := &pb.KVPair{}
+		if !opts.IterNoKeys {
+			pair.Key = object.Key
+			pair.Namespace = object.Namespace
+		}
+
+		if !opts.IterNoValues {
+			pair.Value = object.Data
+		}
+
+		if opts.ReturnMeta {
+			pair.Meta = returnMeta(object)
+		}
+
+		out.Values = append(out.Values, pair)
+	}
+
+	if err = iter.Error(); err != nil {
+		log.Error().Err(err).Str("namespace", in.Namespace).Msg("could not iterate")
+		return nil, status.Errorf(codes.FailedPrecondition, "iteration failure: %s", err)
+	}
+
+	// Check if there is a next page cursor
+	if len(cursor.NextKey) != 0 {
+		if out.NextPageToken, err = cursor.Dump(); err != nil {
+			log.Error().Err(err).Str("namespace", in.Namespace).Msg("could not serialize next page token")
+			return nil, status.Error(codes.FailedPrecondition, "could not serialize next page token")
+		}
+	}
+
+	// Request complete
```

Comment:
Just want to check that in our Prometheus integration this is where we'd want to increment the Iter counter and not anywhere else before this?

Reply:
I believe so, yes. I think of the `Iter` as a single read even though it's of multiple keys since the read is across a snapshot; therefore when the scan is complete, that's the time to instrument it (as opposed to each individual key fetch). If we do server-side statefulness for pagination then we might want to create an additional Prometheus metric for the number of open page cursors.